sdk/program/PROJECT.md

```diff
@@ -0,0 +1,138 @@
+
+The issue for on-chain developers has been brought up or debated in issues like:
+
+* https://github.com/solana-labs/solana/issues/32503
+* https://github.com/solana-labs/solana/issues/32625
+* https://github.com/anza-xyz/agave/issues/20
+
+To give on-chain program developers a better experience, it's time to define
+some priorities and a plan to make things better.
+
+## Concepts for the Program SDK
+
+Here are the priorities for the future of the SDK, in order of importance.
+
+1. Secure
+2. Respect semantic versioning
+3. On-chain program developers come first
+4. Possible to build a new program quickly and easily
+5. Sensible defaults, but allow customization
+6. Minimal dependencies
+7. Kitchen sinks only re-export functionality
+8. Default to new crates or files to bundle functionality
+9. If new functionality can't be done in a new crate or file, use features
+10. Easy to publish
+11. Easy for Agave devs to make changes
+
+## Project Plan Phase 1: solana_program updates
+
+### Break it up: if it can reasonably be used independently, it gets its own crate
+
+As one approach to address point number 4 of easily allowing customization, we
+should avoid bundling too much functionality in solana-program, and instead
+break it up sensibly. This way, someone can easily drop in a custom entrypoint
+without pulling in the default during build time.
+
+This also helps address point number 8 of avoiding kitchen sinks. By putting
+only related functionality into separate crates, there is no temptation for
+developers to "just throw it into the SDK." Instead, they'll have to consider
+the best place for their changes.
+
+When breaking up solana-program, we will take it in phases. First, we'll extract
+functionality required for program clients, which typically involve creating
+instructions and deserializing types. Next, we will actually create program
+client crates for the native programs. And finally, we will extract functionality
+used for writing on-chain programs.
+
+Here's a more specific breakdown:
+
+* Add crates required for program clients:
+  * borsh
+  * clock
+  * example-mocks
+  * hash
+  * instruction
+  * message
+  * program-error
+  * pubkey
+  * serde-varint
+  * sysvar
+* Those crates allow us to make program client crates:
+  * address lookup table
+  * bpf loader upgradeable
+  * feature
+  * loader v4
+  * stake
+  * system
+  * vote
+* Next, focus on crates for writing programs:
+  * account-info
+  * entrypoint
+  * program-pack
+  * program-syscalls
+  * stable-layout
+* Finally, do all the rest:
+  * native-token
+  * serialize-utils
+  * wasm
+* Deprecate any bits that make sense
+* `solana_program` becomes a crate with only re-exports and deprecated code
+
+### Add crate features
+
+After solana-program has been broken up, the next bit is to address point number
+7 and minimize dependencies. This will be done 
+
+* Serialization: serde / borsh / bincode / bytemuck optional
```

Comment:
This needs to be done when/before each new crate is pulled out. If you make a new crate where serde is not optional, then later make it optional, that's a breaking change (even if you make it a default feature, what if the dependent could already be setting `default-features = false`). So it makes sense to handle that when making the new crate.

Reply:
That's an excellent point, thanks

Reply:
Also I think making proc-macro usage optional for most of these new crates is achievable and worthwhile, since proc-macro usage impacts build times quite severely. 

The one part I'm not sure about is stuff that uses `CloneZeroed` - this can be replaced with what it expands to but that does make it harder to maintain. But then again very few types derive CloneZeroed and they don't look like they're gonna change(?)